pkg/controllers/workgenerator/reconciler.go

```diff
@@ -0,0 +1,299 @@
+}
+
+// upsertWork creates or updates the work for the corresponding resource snapshot
+func (r *Reconciler) upsertWork(ctx context.Context, work *workv1alpha1.Work, resourceSnapshot *fleetv1beta1.ClusterResourceSnapshot, resourceBinding *fleetv1beta1.ClusterResourceBinding) error {
+	needCreate := false
+	if work == nil {
+		needCreate = true
+		work = &workv1alpha1.Work{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      resourceSnapshot.Name, // work name same as the corresponding resource snapshot name
+				Namespace: fmt.Sprintf(utils.NamespaceNameFormat, resourceBinding.Spec.TargetCluster),
+				Labels: map[string]string{
+					fleetv1beta1.ParentBindingLabel: resourceBinding.Name,
+					fleetv1beta1.CRPTrackingLabel:   resourceBinding.Labels[fleetv1beta1.CRPTrackingLabel],
+				},
+				OwnerReferences: []metav1.OwnerReference{
+					{
+						APIVersion:         fleetv1beta1.GroupVersion.String(),
+						Kind:               resourceBinding.Kind,
+						Name:               resourceBinding.Name,
+						UID:                resourceBinding.UID,
+						BlockOwnerDeletion: pointer.Bool(true), // make sure that the k8s will call work delete when the binding is deleted
+					},
+				},
+			},
+		}
+	}
+	workObj := klog.KObj(work)
+	resourceBindingObj := klog.KObj(resourceBinding)
+	// reset the manifest list regardless
+	work.Spec.Workload.Manifests = make([]workv1alpha1.Manifest, 0)
+	for _, selectedResource := range resourceSnapshot.Spec.SelectedResources {
+		work.Spec.Workload.Manifests = append(work.Spec.Workload.Manifests, workv1alpha1.Manifest{
+			RawExtension: selectedResource.RawExtension,
+		})
+	}
+
+	// upsert the work
+	if needCreate {
+		if err := r.Client.Create(ctx, work); err != nil {
+			klog.ErrorS(err, "failed to create the work associated with the resourceBinding", "resourceBinding", resourceBindingObj,
+				"work", workObj)
+			return controller.NewAPIServerError(false, err)
+		}
+	} else {
+		if err := r.Client.Update(ctx, work); err != nil {
+			klog.ErrorS(err, "failed to update the work associated with the resourceBinding", "resourceBinding", resourceBindingObj,
+				"work", workObj)
+			return controller.NewAPIServerError(false, err)
+		}
+	}
+
+	klog.V(4).InfoS("upsert the work associated with the resourceBinding", "create", needCreate, "resourceBinding", resourceBindingObj,
+		"work", workObj)
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager.
+// It watches binding events and also delete events for work.
+func (r *Reconciler) SetupWithManager(mgr ctrl.Manager) error {
+	r.recorder = mgr.GetEventRecorderFor("work generator")
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&fleetv1beta1.ClusterResourceBinding{}, builder.WithPredicates(predicate.GenerationChangedPredicate{})).
+		Watches(&source.Kind{Type: &workv1alpha1.Work{}}, &handler.Funcs{
+			// we only care about work delete
+			DeleteFunc: func(evt event.DeleteEvent, queue workqueue.RateLimitingInterface) {
```

Comment:
Could you please add a comment to explain why?

If i understand it correctly, customer cannot create/update/delete the resources in the hub cluster reserved namespaces, so we don't need to sync the work spec.
Here we watch the work delete event to handle the binding deletion.